src/engine/LazyGroupBy.cpp

```diff
@@ -0,0 +1,310 @@
+//   Copyright 2024, University of Freiburg,
+//   Chair of Algorithms and Data Structures.
+//   Author: Robin Textor-Falconi <textorr@informatik.uni-freiburg.de>
+
+#include "LazyGroupBy.h"
+
+#include "engine/sparqlExpressions/LiteralExpression.h"
+#include "global/Constants.h"
+
+template <size_t NUM_AGGREGATED_COLS>
+LazyGroupBy<NUM_AGGREGATED_COLS>::LazyGroupBy(
+    LocalVocab& localVocab,
+    std::vector<GroupBy::HashMapAliasInformation> aggregateAliases,
+    const std::vector<size_t>& groupByCols)
+    : localVocab_{localVocab},
+      aggregateAliases_{std::move(aggregateAliases)},
+      groupByCols_{groupByCols},
+      aggregationData_{initializeAggregationData(aggregateAliases_)},
+      resetAggregationData_{initializeResetAggregationData(aggregationData_)} {
+  for (auto& alias : aggregateAliases_) {
+    for (auto& aggregate : alias.aggregateInfo_) {
+      calculationFunctions_.at(aggregate.aggregateDataIndex_) = std::visit(
+          [this](const auto& aggregateData) -> std::function<ValueId()> {
+            return [&] { return aggregateData.calculateResult(&localVocab_); };
+          },
+          aggregationData_.at(aggregate.aggregateDataIndex_));
+      addToAggregateFunctions_.at(aggregate.aggregateDataIndex_) = std::visit(
+          [](auto& aggregateData)
+              -> std::function<void(
+                  const std::variant<ValueId, LocalVocabEntry>&,
+                  const sparqlExpression::EvaluationContext*)> {
+            return [&](const std::variant<ValueId, LocalVocabEntry>& id,
+                       const sparqlExpression::EvaluationContext* ctx) {
+              return aggregateData.addValue(id, ctx);
+            };
+          },
+          aggregationData_.at(aggregate.aggregateDataIndex_));
+    }
+  }
+}
+
+// _____________________________________________________________________________
+template <size_t NUM_AGGREGATED_COLS>
+std::function<void()>
+LazyGroupBy<NUM_AGGREGATED_COLS>::initializeResetAggregationData(
+    std::array<GroupBy::AggregationData, NUM_AGGREGATED_COLS>&
+        aggregationData) {
+  return std::apply(
+      [](auto&... variants) {
+        // Prevent exploding compilation times as the number of aggregated
+        // columns increased compilation times exponentially
```

Comment:
This needs closer observation to see wrapping function calls in `std::function` makes a measurable difference, or if the compiler just optimizes those away